src/dftbp/geoopt/steepdesc.F90

```diff
@@ -7,199 +7,107 @@
+  end type TSteepdescInput
 
-    !> Weights for the coordinates
-    real(dp), allocatable :: weight(:)
-
-    !> Tolerance criteria for convergence
-    real(dp) :: tolerance
-
-    !> Maximal displacement along one coordinate in one step
-    real(dp) :: maxDisp
 
-    !> If CG converged
-    logical :: tConverged
+  !> Steepest descent optimization driver
+  type, extends(TOptimizer) :: TSteepdesc
 
-    !> If object is initialized
-    logical :: tInitialized
-  end type TSteepDesc
-
-
-  !> Creates SD instance
-  interface init
-    module procedure SteepDesc_init
-  end interface
+    !> Number of variables to optimize
+    integer :: nVar
 
+    !> Weights of gradient components
+    real(dp), allocatable :: weight(:)
 
-  !> Resets the SD instance
-  interface reset
-    module procedure SteepDesc_reset
-  end interface
+  contains
 
+    !> Calculate displacement from gradient
+    procedure :: step
 
-  !> Passes calculated gradient to the minimizer and returns a new point
-  interface next
-    module procedure SteepDesc_next
-  end interface
+    !> Reset optimizer
+    procedure :: reset
 
-  public :: TSteepDesc
-  public :: init, reset, next
+  end type TSteepdesc
 
 contains
 
 
-  !> Initialises a steepest descent instance
-  subroutine SteepDesc_init(this, nElem, tol, maxDisp, weight)
-
-    !> Steepest descent instance on exit
-    type(TSteepDesc), intent(out) :: this
-
-    !> Nr. of elements in the vectors
-    integer, intent(in) :: nElem
-
-    !> Tolerance for the gradient
-    real(dp), intent(in) :: tol
-
-    !> Maximal displacement in one element in one step
-    real(dp), intent(in) :: maxDisp
-
-    !> The weights of the gradient components
-    real(dp), intent(in) :: weight(:)
-
-    @:ASSERT(nElem > 0)
-    @:ASSERT(tol > 0.0_dp)
-    @:ASSERT(maxDisp > 0.0_dp)
-    @:ASSERT(size(weight) == nElem)
+  !> Creates new steepest descent optimization driver.
+  subroutine TSteepdesc_init(this, input, nVar, mass)
 
-    this%nElem = nElem
-    this%tolerance = tol
-    this%maxDisp = maxDisp
-    allocate(this%weight(nElem))
-    this%weight(:) = weight(:)
-    allocate(this%xOld(nElem))
-    this%tInitialized = .false.
+    !> Instance of the optimizer
+    type(TSteepdesc), intent(out) :: this
 
-  end subroutine SteepDesc_init
+    !> Input for the steepest descent optimizer
+    type(TSteepdescInput), intent(in) :: input
 
+    !> Number of variables to optimize
+    integer, intent(in) :: nVar
 
-  !> Resets CG minimizer
-  subroutine SteepDesc_reset(this, x0)
+    !> Mass of atoms that are allowed to move
+    real(dp), intent(in) :: mass(:)
 
-    !> minimizer object
-    type(TSteepDesc), intent(inout) :: this
+    @:ASSERT(nVar > 0)
+    @:ASSERT(size(mass) * 3 == nVar)
 
-    !> Point to start from
-    real(dp), intent(in) :: x0(:)
+    this%nVar = nVar
+    allocate(this%weight(nVar))
+    this%weight(:) = 0.5_dp * input%stepSize**2 / reshape(spread(mass, 1, 3), [nVar])
```

Comment:
Any reason not to make the StepSize**2 / mass the optimizer argument? It's giving a step distance as length/force, but actually what probably beter is a quantity dimensioned as length/(dE/dx). There is also a question as to why scaling the coordinate system by inverse mass would be a better choice than mass or no scaling at all (the inverse mass means H atoms move further than heavier atoms in each step).

Reply:
Possibly the simplest thing would be to set the compliance (i.e. inverse of stiffness) for the driver (i.e. proportionality between displacement and force). This quantity is dimensioned as length / force, so would  not be in mass weighted coordinates.

Reply:
This would effectively be a single scalar, scaling the gradients right? Or a 1d array of weights?

Reply:
Switched to single scaling factor.